plugin/pkg/shoot/exposureclass/admission.go

```diff
@@ -0,0 +1,210 @@
+	externalcoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	corev1alphalisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1alpha1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+)
+
+const (
+	// PluginName is the name of this admission plugin.
+	PluginName = "ShootExposureClass"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return New()
+	})
+}
+
+// ExposureClass contains listers and admission handler.
+type ExposureClass struct {
+	*admission.Handler
+
+	exposureClassLister corev1alphalisters.ExposureClassLister
+	readyFunc           admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsExternalCoreInformerFactory(&ExposureClass{})
+
+	readyFuncs []admission.ReadyFunc
+)
+
+// New creates a new ExposureClass admission plugin.
+func New() (*ExposureClass, error) {
+	return &ExposureClass{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (e *ExposureClass) AssignReadyFunc(f admission.ReadyFunc) {
+	e.readyFunc = f
+	e.SetReadyFunc(f)
+}
+
+// SetExternalCoreInformerFactory sets the external garden core informer factory.
+func (e *ExposureClass) SetExternalCoreInformerFactory(f externalcoreinformers.SharedInformerFactory) {
+	exposureClassInformer := f.Core().V1alpha1().ExposureClasses()
+	e.exposureClassLister = exposureClassInformer.Lister()
+
+	readyFuncs = append(readyFuncs, exposureClassInformer.Informer().HasSynced)
+}
+
+func (e *ExposureClass) waitUntilReady(attrs admission.Attributes) error {
+	// Wait until the caches have been synced
+	if e.readyFunc == nil {
+		e.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+
+	if !e.WaitForReady() {
+		return admission.NewForbidden(attrs, errors.New("not yet ready to handle request"))
+	}
+
+	return nil
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (e *ExposureClass) ValidateInitialization() error {
+	if e.exposureClassLister == nil {
+		return errors.New("missing ExposureClass lister")
+	}
+	return nil
+}
+
+// Admit unite the seed selector and/or tolerations of a Shoot resource
+// with the ones from the referenced ExposureClass.
+func (e *ExposureClass) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	if err := e.waitUntilReady(a); err != nil {
+		return fmt.Errorf("err while waiting for ready %v", err)
+	}
+
+	if a.GetKind().GroupKind() != core.Kind("Shoot") {
+		return nil
+	}
+
+	if a.GetOperation() != admission.Create {
+		return nil
+	}
```

Comment:
You can make this check obsolete by adapting https://github.com/gardener/gardener/pull/4243/files#diff-299d23f0ced7d6667116c2e458908c621edabce15367f2f878d24f06451350afR65

Reply:
done